fastparquet/core.py

```diff
@@ -273,9 +270,17 @@ def read_row_group(file, rg, columns, categories, schema_helper, cats):
             continue
 
         use = name in categories if categories is not None else False
-        s = read_col(column, schema_helper, file, use_cat=use)
+        s = read_col(column, schema_helper, file, use_cat=use,
+                     selfmade=selfmade)
         out[name] = s
-    out = pd.DataFrame(out, columns=columns)
+    # TODO: out is a dict of numpy arrays and maybe Category arrays;
+    # could optionally return here before dataframe construction
```

Comment:
Perhaps have a separate function, `read_row_group_arrays` that contains most logic and gets called by `read_row_group`?